app/controllers/incomeSources/manage/ManageObligationsController.scala

```diff
@@ -0,0 +1,238 @@
+            handleRequest(
+              mode = SelfEmployment,
+              isAgent = true,
+              taxYear,
+              changeTo,
+              id
+            )
+        }
+  }
+
+  def showUKProperty(changeTo: String, taxYear: String): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(
+        mode = UkProperty,
+        isAgent = false,
+        taxYear,
+        changeTo,
+        ""
+      )
+  }
+
+  def showAgentUKProperty(changeTo: String, taxYear: String): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+          implicit mtdItUser =>
+            handleRequest(
+              mode = UkProperty,
+              isAgent = true,
+              taxYear,
+              changeTo,
+              ""
+            )
+        }
+  }
+
+  def showForeignProperty(changeTo: String, taxYear: String): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(
+        mode = ForeignProperty,
+        isAgent = false,
+        taxYear,
+        changeTo,
+        ""
+      )
+  }
+
+  def showAgentForeignProperty(changeTo: String, taxYear: String): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+          implicit mtdItUser =>
+            handleRequest(
+              mode = ForeignProperty,
+              isAgent = true,
+              taxYear,
+              changeTo,
+              ""
+            )
+        }
+  }
+
+  def handleRequest(mode: Mode, isAgent: Boolean, taxYear: String, changeTo: String, incomeSourceId: String)(implicit user: MtdItUser[_], hc: HeaderCarrier): Future[Result] = {
+    if (isDisabled(IncomeSources)) {
+      if (isAgent) Future.successful(Redirect(controllers.routes.HomeController.showAgent))
+      else Future.successful(Redirect(controllers.routes.HomeController.show()))
+    }
+    else {
+      val backUrl: String = if (isAgent) controllers.incomeSources.manage.routes.ManageConfirmController.showAgent().url else controllers.incomeSources.manage.routes.ManageConfirmController.show().url
+      val postUrl: Call = if (isAgent) controllers.incomeSources.manage.routes.ManageObligationsController.agentSubmit() else controllers.incomeSources.manage.routes.ManageObligationsController.submit()
+
+
+      if (mode == SelfEmployment && !user.incomeSources.businesses.exists(x => x.incomeSourceId.contains(incomeSourceId))) {
+        showError(isAgent, s"unable to find incomeSource by id: $incomeSourceId")
+      }
+      else {
+        val addedBusinessName: String = mode match {
```

Comment:
Ideally we should reduce number of lines in this method as it about 40 lines, which is okish ..
Extracting method to return addedBusinessName would be best candidate for this simplification;